ionoscloud/resource_server.go

```diff
@@ -910,87 +914,107 @@ func resourceServerUpdate(ctx context.Context, d *schema.ResourceData, meta inte
 					}
 				}
 				if ips != nil && len(ips) > 0 {
-					properties.Ips = &ips
+					nicProperties.Ips = &ips
 				}
 			}
 		}
 
 		dhcp := d.Get("nic.0.dhcp").(bool)
 		fwRule := d.Get("nic.0.firewall_active").(bool)
-		properties.Dhcp = &dhcp
-		properties.FirewallActive = &fwRule
+		nicProperties.Dhcp = &dhcp
+		nicProperties.FirewallActive = &fwRule
 
 		if v, ok := d.GetOk("nic.0.firewall_type"); ok {
 			vStr := v.(string)
-			properties.FirewallType = &vStr
+			nicProperties.FirewallType = &vStr
 		}
-
-		if d.HasChange("nic.0.firewall") {
-
-			firewallId := d.Get("firewallrule_id").(string)
-			update := true
-			if firewallId == "" {
-				update = false
+		firstNicFirewallPath := "nic.0.firewall"
+		if d.HasChange(firstNicFirewallPath) {
+			oldValues, newValues := d.GetChange(firstNicFirewallPath)
+			oldFwSlice, newFwSlice, err := GetChangesInFirewallRules(oldValues.([]interface{}), newValues.([]interface{}))
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("could not get changes for FirewallRules %w", err))
 			}
 
-			firewall, diags := getFirewallData(d, "nic.0.firewall.0.", update)
-			if diags != nil {
-				return diags
-			}
+			firewallId := d.Get("firewallrule_id").(string)
+			firewallRuleIds := d.Get("firewallrule_ids").([]interface{})
+			fwRuleIdsString := convertSlice(firewallRuleIds)
 
-			_, apiResponse, err := client.FirewallRulesApi.DatacentersServersNicsFirewallrulesFindById(ctx, dcId, *server.Id, *nic.Id, firewallId).Execute()
-			logApiRequestTime(apiResponse)
+			if nic != nil && nic.Id != nil {
+				fs := FirewallService{client: client, meta: meta, schemaData: d}
 
-			if err != nil {
-				if !httpNotFound(apiResponse) {
-					diags := diag.FromErr(fmt.Errorf("error occured at checking existance of firewall %s %s", firewallId, err))
-					return diags
-				} else if httpNotFound(apiResponse) {
-					diags := diag.FromErr(fmt.Errorf("firewall does not exist %s", firewallId))
-					return diags
+				firewalls, err := fs.firewallsGet(ctx, dcId, *server.Id, *nic.Id, 1)
+				if err != nil {
+					return diag.FromErr(fmt.Errorf("an error occurred while fetching firewall rules: for dcId: %s server_id: %s nic_id %s `%w`", dcId, *server.Id, *nic.Id, err))
+				}
+				//delete old rules
+				for _, ruleProp := range oldFwSlice {
+					for _, item := range firewalls {
```

Comment:
@cristiGuranIonos this naming is a little bit confusing and this is hard to read, I think we should use something else instead of `item`, I know that this is a tricky context and it will still be complicated no matter the naming, but maybe we can figure something out.